service/src/main/kotlin/fi/espoo/evaka/invoicing/controller/FinanceBasicsController.kt

```diff
@@ -158,6 +160,50 @@ class FinanceBasicsController(
             .also { Audit.FinanceBasicsVoucherValuesRead.log() }
     }
 
+    @PostMapping("/voucher-values")
+    fun createVoucherValue(
+        db: Database,
+        user: AuthenticatedUser,
+        clock: EvakaClock,
+        @RequestBody body: ServiceNeedOptionVoucherValueRange
+    ) {
+        val id =
+            db.connect { dbc ->
+                dbc.transaction { tx ->
+                    accessControl.requirePermissionFor(
+                        tx,
+                        user,
+                        clock,
+                        Action.Global.CREATE_VOUCHER_VALUE
+                    )
+
+                    val serviceNeedOption =
+                        tx.getServiceNeedOptions()
+                            .filter { it.id == body.serviceNeedOptionId }
+                            .firstOrNull()
+                    if (serviceNeedOption == null)
+                        throw BadRequest("Invalid service need option ID")
+
+                    val currentVoucherValues =
+                        tx.getVoucherValuesByServiceNeedOption()[body.serviceNeedOptionId]!!
+
+                    val latest = currentVoucherValues.maxByOrNull { it.voucherValues.range.start }
+
+                    if (latest != null && latest.voucherValues.range.end == null)
+                        tx.updateVoucherValueEndDate(latest.id, body.range.start)
+
+                    tx.insertNewVoucherValue(body).also {
+                        asyncJobRunner.plan(
+                            tx,
+                            listOf(AsyncJob.NotifyFeeThresholdsUpdated(body.range)),
```

Comment:
might use `DateRange(body.range.start, null)` because the above `updateVoucherValueEndDate` may have affected time after `body.range`

Reply:
Good point, fixed.